rules/onchain_budget.go

```diff
@@ -0,0 +1,785 @@
+type OnChainBudget struct {
+	AbsoluteAmtSats uint64 `json:"absolute_amt_sats"`
+	MaxSatPerVByte  uint64 `json:"max_sat_per_v_byte"`
+}
+
+// VerifySane checks that the value of the values is ok given the min and max
+// allowed values.
+//
+// NOTE: this is part of the Values interface.
+func (o *OnChainBudget) VerifySane(minVal, _ Values) error {
+	minOCB, ok := minVal.(*OnChainBudget)
+	if !ok {
+		return fmt.Errorf("min value is not of type OnChainBudget")
+	}
+
+	if o.AbsoluteAmtSats < minOCB.AbsoluteAmtSats {
+		return fmt.Errorf("on-chain budget is below the minimum " +
+			"required amount")
+	}
+
+	if o.MaxSatPerVByte < minOCB.MaxSatPerVByte {
+		return fmt.Errorf("on-chain fee rate limit is below the " +
+			"minimum required amount")
+	}
+
+	return nil
+}
+
+// RuleName returns the name of the rule that these values are to be used with.
+//
+// NOTE: this is part of the Values interface.
+func (o *OnChainBudget) RuleName() string {
+	return OnChainBudgetName
+}
+
+// ToProto converts the rule Values to the litrpc counterpart.
+//
+// NOTE: this is part of the Values interface.
+func (o *OnChainBudget) ToProto() *litrpc.RuleValue {
+	return &litrpc.RuleValue{
+		Value: &litrpc.RuleValue_OnChainBudget{
+			OnChainBudget: &litrpc.OnChainBudget{
+				AbsoluteAmtSats: o.AbsoluteAmtSats,
+				MaxSatPerVByte:  o.MaxSatPerVByte,
+			},
+		},
+	}
+}
+
+// PseudoToReal attempts to convert any appropriate pseudo fields in the rule
+// Values to their corresponding real values. It uses the passed PrivacyMapDB to
+// find the real values. This is a no-op for the OnChainBudget rule.
+//
+// NOTE: this is part of the Values interface.
+func (o *OnChainBudget) PseudoToReal(_ firewalldb.PrivacyMapDB,
+	_ session.PrivacyFlags) (Values, error) {
+
+	return o, nil
+}
+
+// RealToPseudo converts the rule Values to a new one that uses pseudo keys,
+// channel IDs, channel points etc. It returns a map of real to pseudo strings
+// that should be persisted. This is a no-op for the OnChainBudget rule.
+//
+// NOTE: this is part of the Values interface.
+func (o *OnChainBudget) RealToPseudo(db firewalldb.PrivacyMapReader,
+	flags session.PrivacyFlags) (Values, map[string]string, error) {
+
+	return o, nil, nil
+}
+
+// checkFeeRate is a helper function that returns an error if the given fee rate
+// exceeds the maximum allowed by the rule that the enforcer is enforcing.
+func (o *OnChainBudgetEnforcer) checkFeeRate(satPerVByte uint64) error {
+	if satPerVByte > o.MaxSatPerVByte {
+		return fmt.Errorf("max fee rate exceeded")
+	}
+	return nil
+}
+
+// formatReqId creates a unique identifier for the request that can be used to
+// associate the action accross request-response handling.
+func formatReqId(lndConnID string, reqID int64) string {
+	return fmt.Sprintf("%s-%s-%d", memoPrefix, lndConnID, reqID)
+}
+
+// removeReqId removes the request ID from the memo if present.
+func removeReqId(memo string) string {
+	parts := strings.Split(memo, ":")
```

Comment:
It might be worth already implementing removal of the memo in by looping the over the parts and to look for the onBudget prefix to avoid that older versions of litd in the future leak extra privacy, and to have a reference how it should be done if we implement useage of the memo field for fetching the reqId in the future.

Reply:
I have implemented that idea and tested the current prefix together with a different one